ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplicationMigrationTool.java

```diff
@@ -0,0 +1,537 @@
+    // entry missing or mismatch(error case), everything should match in sequence.
+    while (srcListing.hasNext() && tgtListing.hasNext()) {
+      LocatedFileStatus sourceFile = srcListing.next();
+
+      // Check if the entry is a filtered entry, if the entry is a filtered entry we can skip comparing and proceed
+      // further, since that entry won't be there in the target, considering the same filter would be passed during
+      // replication.
+      if (filtersPattern != null && !isCopied(sourceFile.getPath(), filtersPattern)) {
+        LOG.info("Entry: {} is filtered.", sourceFile.getPath());
+        continue;
+      }
+      LocatedFileStatus tgtFile = tgtListing.next();
+      LOG.info("Comparing {} and {}", sourceFile.getPath(), tgtFile.getPath());
+      String sourcePath = Path.getPathWithoutSchemeAndAuthority(sourceFile.getPath()).toString()
+          .replaceFirst(Path.getPathWithoutSchemeAndAuthority(srcDir).toString(), "/");
+      String targetPath = Path.getPathWithoutSchemeAndAuthority(tgtFile.getPath()).toString()
+          .replaceFirst(Path.getPathWithoutSchemeAndAuthority(tgtDir).toString(), "/");
+      if (!sourcePath.equals(targetPath)) {
+        System.err.println(
+            "Entries mismatch between source: " + srcDir + " and target: " + tgtDir + " for " + "sourceFile: "
+                + sourceFile.getPath() + " at target with " + tgtFile.getPath() + " Mismatched subPaths: " + "source: "
+                + sourcePath + " and target: " + targetPath + "Either Source Or Target has an extra/less files.");
+        // Get the response from all the checksum tasks scheduled, in order to prevent the client being terminating
+        // abruptly.
+        validateChecksumStatus(srcDir, tgtDir, futures);
+        // Entries have mismatched, the dir needs to copied again, return false, and print the source dir, so it can
+        // be copied entirely.
+        return false;
+      }
+      if (sourceFile.getLen() != tgtFile.getLen()) {
+        System.err.println("File Size mismatch in source directory " + srcDir + " and target directory " + tgtDir + " "
+            + "for source: " + sourceFile.getPath() + " and target " + tgtFile.getPath());
+        // Try further if only some files have been modified, trying out if there are no addition/deletion of files in
+        // directory on moving ahead.
+        response = false;
+      }
+      LOG.debug("Length matched for {} and {}", sourceFile.getPath(), tgtFile.getPath());
+      // Verify checksum of the file.
+      if (verifyChecksum) {
+        ChecksumVerifier checksumVerifier =
+            new ChecksumVerifier(sourceFile.getPath(), tgtFile.getPath(), srcFs, tgtFs, srcDir, tgtDir);
+        Future<Boolean> future = threadPoolChecksum.submit(checksumVerifier);
+        futures.add(future);
+      }
+    }
+    // Check if entries are left in the target.
+    if (tgtListing.hasNext()) {
```

Comment:
Maybe we can expose one option for customer whether to fail or not if target has extra entries.

Reply:
That will not work. In general the source and target listing is being compared above. if there is a mismatch above, it will break there itself.

This code is just checking a corner case, like both source & target listing got matched. but in the end target still had entries, which would land up in the end in the sorted order.

Like
/src has /1 /4 /6
and 
/trg has /1 /4 /6 /8

This one is to catch /8 since till /6 everything will match in above code only. If there is mismatch or extra entry in before /6 say target has /2 also, in that case, the code will break above only.